openmls/src/group/tests/test_proposal_validation.rs

```diff
@@ -98,61 +152,191 @@ fn test_valsem100(ciphersuite: &'static Ciphersuite, backend: &impl OpenMlsCrypt
+                    .expect("failed to generate signature keypair");
+
+            bob_signature_keypair = shared_signature_keypair.clone();
+            charlie_signature_keypair = shared_signature_keypair.clone();
+        } else {
+            bob_signature_keypair = SignatureKeypair::new(ciphersuite.signature_scheme(), backend)
+                .expect("failed to generate signature keypair");
+            charlie_signature_keypair =
+                SignatureKeypair::new(ciphersuite.signature_scheme(), backend)
+                    .expect("failed to generate signature keypair");
+        }
+
+        let bob_credential_bundle =
+            CredentialBundle::from_parts("Bob".into(), bob_signature_keypair);
+        let charlie_credential_bundle =
+            CredentialBundle::from_parts("Charlie".into(), charlie_signature_keypair);
+
+        let bob_key_package_bundle = KeyPackageBundle::new(
+            &[ciphersuite.name()],
+            &bob_credential_bundle,
+            backend,
+            vec![],
+        )
+        .expect("failed to generate key package");
+        let bob_key_package = bob_key_package_bundle.key_package().clone();
+        let charlie_key_package_bundle = KeyPackageBundle::new(
+            &[ciphersuite.name()],
+            &charlie_credential_bundle,
+            backend,
+            vec![],
+        )
+        .expect("failed to generate key package");
+        let charlie_key_package = charlie_key_package_bundle.key_package().clone();
+
+        // 2. Create a group and try to create a commit to add Bob and Charlie
+        let res = create_commit_to_add_bob_and_charlie(
+            alice_credential_bundle,
+            alice_key_package_bundle,
+            bob_key_package,
+            charlie_key_package,
+            ciphersuite,
+            backend,
+        );
+
+        if bob_and_charlie_share_keys {
+            // Negative Case: we should output an error
+            let err =
+                res.expect_err("Created commit when the proposals have a same signature key!");
+            assert_eq!(
+                err,
+                CoreGroupError::ProposalValidationError(
+                    ProposalValidationError::DuplicateSignatureKeyAddProposal
+                )
+            );
+        } else {
+            // Positive Case: we should succeed
+            let _ = res.expect("Failed to create commit with different signature keypairs!");
+        }
+    }
+
+    // TODO #525: Add test for incoming proposals.
+}
+
+/// ValSem102:
+/// Add Proposal:
+/// HPKE init key in proposals must be unique among proposals
+#[apply(ciphersuites_and_backends)]
+fn test_valsem102(ciphersuite: &'static Ciphersuite, backend: &impl OpenMlsCryptoProvider) {
+    for bob_and_charlie_share_hpke_init_keys in [
+        true,  // Negative Case: Bob and Charlie have same HPKE init key
+        false, // Positive Case: Bob and Charlie have different HPKE init keys
+    ] {
```

Comment:
Not strictly necessary, but things become a little rustier by using enums because you can later `match` on them.

Example:

```rust
enum TestCase {
    PositiveDifferentKey,
    NegativeSameKey,
}

match test_case {
    TestCase::PositiveDifferentKey => {...},
    TestCase::NegativeSameKey => {...},
}
```

Reply:
Thank you for the suggestion. :)